foyer-storage/src/large/scanner.rs

```diff
@@ -12,166 +12,95 @@
-        );
-        let end = std::cmp::min(self.region.size(), end);
-
-        let read_len = end - self.offset as usize;
-        assert!(bits::is_aligned(self.region.align(), read_len));
-        assert!(read_len >= len);
-
-        let buffer = self.region.read(self.offset, read_len).await?.freeze();
-        self.buffer = buffer;
-
-        let start = (offset - self.offset) as usize;
-        let end = start + len;
-        Ok(&self.buffer[start..end])
-    }
-}
-
 #[derive(Debug)]
 pub struct RegionScanner {
     region: Region,
     offset: u64,
-    cache: CachedRegionReader,
-    _metrics: Arc<Metrics>,
 }
 
 impl RegionScanner {
-    pub fn new(region: Region, metrics: Arc<Metrics>) -> Self {
-        let cache = CachedRegionReader::new(region.clone());
-        Self {
-            region,
-            offset: 0,
-            cache,
-            _metrics: metrics,
-        }
+    pub fn new(region: Region) -> Self {
+        Self { region, offset: 0 }
     }
 
-    async fn current(&mut self) -> Result<Option<EntryHeader>> {
-        strict_assert!(bits::is_aligned(self.region.align() as u64, self.offset));
-
-        if self.offset as usize >= self.region.size() {
-            // reach region EOF
-            return Ok(None);
-        }
-
-        // load entry header buf
-        let buf = self.cache.read(self.offset, EntryHeader::serialized_len()).await?;
-
-        if buf.len() < EntryHeader::serialized_len() {
+    pub async fn next(&mut self) -> Result<Option<Vec<EntryInfo>>> {
+        if self.offset as usize + PAGE > self.region.size() {
             return Ok(None);
         }
 
-        let res = EntryHeader::read(buf).ok();
-
-        Ok(res)
-    }
-
-    async fn step(&mut self, header: &EntryHeader) {
-        let aligned = bits::align_up(self.region.align(), header.entry_len());
-        self.offset += aligned as u64;
-    }
-
-    fn info(&self, header: &EntryHeader) -> EntryInfo {
-        EntryInfo {
-            hash: header.hash,
-            sequence: header.sequence,
-            addr: EntryAddress {
-                region: self.region.id(),
-                offset: self.offset as _,
-                len: header.entry_len() as _,
-                sequence: header.sequence,
-            },
-        }
-    }
-
-    pub async fn next(&mut self) -> Result<Option<EntryInfo>> {
-        let header = match self.current().await {
-            Ok(Some(header)) => header,
-            Ok(None) => return Ok(None),
-            Err(e) => return Err(e),
+        let page = IoBuffer::new(PAGE);
+        let (page, res) = self.region.read(page, self.offset).await;
+        res?;
+        let blob = match BlobReader::read(page) {
+            Ok(blob) => blob,
+            Err(e) => {
+                tracing::trace!(
+                    ?e,
+                    region_id = self.region.id(),
+                    "[region blob scanner]: error to parse a blob, skip the remaining region"
```

Comment:
Does it mean this is backward incompatible?

Reply:
It is incompatible. But it will be fine as a cache system. Besides the previous recovery is barely used.